ark/phenotyping/pixel_consensus_cluster.R

```diff
@@ -100,11 +109,46 @@ for (batchStart in seq(1, length(fovs), batchSize)) {
     batchEnd <- min(batchStart + batchSize - 1, length(fovs))
 
     # run the multithreaded batch process for mapping to SOM labels and saving
-    foreach(
+    fovStatuses <- foreach(
         i=batchStart:batchEnd,
-        .combine='c'
+        .combine=rbind
     ) %dopar% {
-        mapConsensusLabels(fovs[i], pixelMatDir, som_to_meta_map)
+        fileName <- paste0(fovs[i], '.feather')
+        matPath <- file.path(pixelMatDir, fileName)
+
+        status <- tryCatch(
+            {
+                fovPixelData <- arrow::read_feather(matPath)
+
+                # assign hierarchical cluster labels
+                fovPixelData$pixel_meta_cluster <- som_to_meta_map[as.character(fovPixelData$pixel_som_cluster)]
+
+                # write data with consensus labels
+                tempPath <- file.path(paste0(pixelMatDir, '_temp'), fileName)
+                arrow::write_feather(as.data.table(fovPixelData), tempPath, compression='uncompressed')
+
+                # this won't be displayed to the user but is used as a helper to break out
+                # in the rare first FOV hang issue
+                print(paste('Done writing fov', fovs[i]))
+                0
+            },
+            error=function(cond) {
+                # this won't be displayed to the user but is used as a helper to break out
+                # in the rare first FOV hang issue
+                print(paste('Error encountered for fov', fovs[i]))
+                1
+            }
+        )
+
+        data.frame(fov=fovs[i], status=status)
+    }
+
+    # report any erroneous feather files
+    for (i in 1:nrow(fovStatuses)) {
+        if (fovStatuses[i, 'status'] == 1) {
+            print(paste("The data for FOV", fovStatuses[i, 'fov'], "has been corrupted, removing"))
```

Comment:
Does this file get removed somewhere? How often do these corruption issues come up now that we've removed the feather overwriting? Is it related to multi-processing, or inherent to writing the files

Reply:
This could happen if the writing process for a FOV in `run_pixel_som.R` terminated early (ex. the kernel crashed). On restart, since the FOV is technically partially written, `run_pixel_som.R` will not check to re-generate this (it would be cumbersome to open up 100s of large FOVs to see which ones were invalid). So this will need to be handled by `pixel_consensus_cluster.R`.

Mostly, this exception handling is in place to prevent overnight crashing from stopping the entire pipeline. That being said, I agree that this issue shouldn't come up frequently with the new directory structure, so I can remove it.

Reply:
@ngreenwald per our discussion today, we'll leave the `try`/`except` logic in.